core/win32/drwinapi/kernel32_proc.c

```diff
@@ -48,8 +48,14 @@ void
 kernel32_redir_init_proc(void)
 {
     PEB *peb = get_own_peb();
+    /* FIXME: i#3633: Fix for Windows 10 1903 where PEB->Fls* is replaced with
+     * PEB->SparePointers and PEB->SpareUlongs. Maybe this code shouldn't even
+     * check for peb->FlsBitmap, as in ntdll_redir.c we are anyhow allocating
+     * own FLS structure globaly and we don't relay at all on Windows FLS
+     * implementation
+     */
```

Comment:
While this does give us a single solution independent of the Windows version, it does mean that if there's some code (say in a Windows system library we've loaded privately for the client) that examines PEB->Fls* we'll hit problems: i.e., we're relying on intercepting *everything* that touches Fls, whereas un-intercepted code has a chance of working if the data was in the same place.  This approach still seems fine since we are intercepting for other reasons, but it may be worth a comment acknowledging the assumption on interception.

Reply:
Comment can be reformulated to fully describe what is (and why) going on.

What bothers me here, is certain scenario which I want to test. If somebody knows answer to next situation would be nice to jump in and spare time in debugging. 

Here is scenario which puzzles me (but I have to verify):
1. lets load ucrtbase.dll from client - private_load. flsIndex will be allocated with redirection, and stored in global var in ucrtbase.dll as a part of ucrtbase!CRT_INIT. 
2. teb->FlsData at this point is changed to private FlsData (priv_fls_data).
3. teb->FlsData[flsIndex] = something_from_ucrtbase.dll
4. Now, maybe through delay load import, this dll is being resolved in original app and API gets called which uses global flsIndex. In this case teb->FlsData (app_fls_data) is original from the app, and crt expects to find valid data there. It should either get wrong data (NULL) or FLS set by some other dll. Wanted/expected data resided inside of a dcontext->client_data->priv_fls_data, and not inside of a dcontext->client_data->app_fls_data;

I will run this test, and post update. Just to be 100% sure that code doesn't run into some weird case which could lead to hard-to-debug crashes.

Reply:
I don't follow: if the app loads ucrtbase, ucrtbase!CRT_INIT should be executed in an app context which will initialize the app's version of the FLS.

Reply:
I don't know, maybe or maybe not depending if the app will be using it or not. One solution would be that teb->FlsData is not replaced, but as redir_Fls*/redir_RtlFls* is applied in Get/SetValue to directly use dcontext->client_info->priv_fls_data; as flsindex will always point to real data. Yes, it could happen that client by using these APIs will obtain FLS data in private dynamorio memory, but there will be no conflict using one globally saved flsindex. 

Apparently this is not a big issue, as nobody noticed it over the years. I would keep it as is, and keep it in mind in case somebody ever runs into it.